### PR TITLE
Arc 1042 Added GitHub client timeout

### PR DESCRIPTION
**Pull request by KAbakumovAtl**
In this PR I've added a feature flag to control GitHub Client Timeout

I've made it a number feature flag so we can test which timeout better (30sec vs 10sec)

### CODE REVIEW

File: src/github/client/github-client.ts [hunk 1]
```diff
@@ -31,15 +32,32 @@ export default class GitHubClient {
 	private readonly githubInstallationId: InstallationId;
 	private readonly logger: Logger;
 
-	constructor(
+	public static async build(githubInstallationId: InstallationId,
+		logger: Logger,
+		appTokenHolder: AppTokenHolder = AppTokenHolder.getInstance()) : Promise<GitHubClient> {
+		const timeout = await numberFlag(NumberFlags.GITHUB_CLIENT_TIMEOUT, 30000);
+		return new GitHubClient(githubInstallationId, logger, appTokenHolder, timeout);
```
Comment (KAbakumovAtl) [inline src/github/client/github-client.ts:39]: I had to make a builder method here, and made the constructor private.

Because constructor can't be async

File: src/github/client/github-client.ts [hunk 2]
```diff
@@ -31,15 +32,32 @@ export default class GitHubClient {
...
-		appTokenHolder: AppTokenHolder = AppTokenHolder.getInstance()
+		appTokenHolder: AppTokenHolder = AppTokenHolder.getInstance(),
+		timeout: number
 	) {
 		this.logger = logger || getLogger("github.client.axios");
-		this.axios = axios.create({
-			baseURL: githubInstallationId.githubBaseUrl
-		});
+
+
```
Comment (KAbakumovAtl) [inline src/github/client/github-client.ts:50]: /nit empty lines

File: src/config/feature-flags.ts [hunk 3]
```diff
@@ -61,6 +65,11 @@ export const booleanFlag = async (flag: BooleanFlags, defaultValue: boolean, jir
 export const stringFlag = async (flag: StringFlags, defaultValue: string, jiraHost?: string): Promise<string> =>
 	String(await getLaunchDarklyValue(flag, defaultValue, jiraHost));
 
+export const numberFlag = async (flag: NumberFlags, defaultValue: number, jiraHost?: string): Promise<number> =>
+	Number(await getLaunchDarklyValue(flag, defaultValue, jiraHost));
+
+
+
```
Comment (KAbakumovAtl) [inline src/config/feature-flags.ts:72]: /nit empty lines

File: src/github/client/interceptors.ts [hunk 4]
```diff
@@ -75,6 +77,12 @@ export const instrumentFailedRequest = (metricName) =>
 export const handleFailedRequest = (logger: Logger) =>
 	(error) => {
 		const response = error.response as AxiosResponse;
+
+		if (response?.status === 408 || error.code === "ECONNABORTED") {
```
Comment (KAbakumovAtl) [inline src/github/client/interceptors.ts:81]: That's the only way to determine if it is a timeout with Axios client (See https://github.com/axios/axios/issues/1174)

File: src/jira/client/axios.ts [hunk 5]
```diff
@@ -133,6 +134,8 @@ const instrumentRequest = (response) => {
 		status: response.status
 	};
 
+	statsd.histogram(metricHttpRequest.jira, requestDurationMs, tags);
+	tags["gsd_histogram"] = RESPONSE_TIME_HISTOGRAM_BUCKETS;
 	statsd.histogram(metricHttpRequest.jira, requestDurationMs, tags);
```
Comment (KAbakumovAtl) [inline src/jira/client/axios.ts:140]: That's from another PR. It will go away from here after it merged

**Comment by KAbakumovAtl** [comment]
- [ ] FD Ticket to fill up go/flag-ticket/FD-29853

File: src/github/client/interceptors.ts (hunk 4 above)
Comment (mboudreau) [reply src/github/client/interceptors.ts:81]: Yes, but no.  This can also be the status/error code for aborting a request from our client side.

You should use the axios specific timeout code of `ETIMEDOUT`.  Upgrade Axios to the latest version (0.25.0) and add a transitional object to the config:
```
transitional: {
    clarifyTimeoutError: true
  },
  ```

This will make axios differentiate between the two.  I believe this will be the default moving forward.

**Review by mboudreau — CHANGES_REQUESTED** [review]
Coding pattern needs to be improved or else we can't change timeout dynamically as well as error handling

File: src/github/client/interceptors.ts (hunk 4 above)
Comment (mboudreau) [reply src/github/client/interceptors.ts:81]: More on this here: https://github.com/axios/axios/blob/master/lib/adapters/xhr.js#L133

There's a lot of talk about this on the project and by using this pattern will help future proof our code for new axios versions.

File: src/github/client/github-client.ts [hunk 6]
```diff
@@ -31,15 +32,32 @@ export default class GitHubClient {
 	private readonly githubInstallationId: InstallationId;
 	private readonly logger: Logger;
 
-	constructor(
+	public static async build(githubInstallationId: InstallationId,
```
[the remark this answers is not in the document]
Comment (KAbakumovAtl) [reply src/github/client/github-client.ts:35]: Our client is not a long-living entity

We create a new client instance when:
1. We process SQS Message
2. Cann get-github-configuration REST API

That mean the timeout will be active only within a scope of a small operation, which usually takes less than a minute
Hence we don't care if our feature flag change won't be applied immediately.

File: src/github/client/interceptors.ts (hunk 4 above)
Comment (KAbakumovAtl) [reply src/github/client/interceptors.ts:81]: Oh great, I thought they didn't implement it.

File: src/github/client/github-client.ts [hunk 7]
```diff
@@ -31,15 +32,32 @@ export default class GitHubClient {
...
+		logger: Logger,
+		appTokenHolder: AppTokenHolder = AppTokenHolder.getInstance()) : Promise<GitHubClient> {
+		const timeout = await numberFlag(NumberFlags.GITHUB_CLIENT_TIMEOUT, 30000);
```
[the remark this answers is not in the document]
Comment (KAbakumovAtl) [reply src/github/client/github-client.ts:38]: In unlikely case if LD is down we want to keep it as if it was rolled out

File: src/github/client/interceptors.ts [hunk 8]
```diff
@@ -18,6 +19,17 @@ export const setRequestStartTime = (config) => {
 	return config;
 };
 
+/**
+ * Sets the timeout to the request based on the github-client-timeout feature flag
+ */
+export const setRequestTimeout = async (config: AxiosRequestConfig): Promise<AxiosRequestConfig> => {
+	const timeout = await numberFlag(NumberFlags.GITHUB_CLIENT_TIMEOUT, 30000);
+	if(timeout) {
```
Comment (mboudreau) [inline src/github/client/interceptors.ts:27]: This will always return true?

File: src/sync/pull-request.ts [hunk 9]
```diff
@@ -46,7 +46,7 @@ export default async function(
 ) {
 	logger.info("Syncing PRs: started");
 	let status: number;
-	let headers: Headers = {};
```
Comment (mboudreau) [inline src/sync/pull-request.ts:49]: Why this change?

File: src/github/client/github-client.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (mboudreau) [reply src/github/client/github-client.ts]: @KAbakumovAtl Oh.  That's not good.  The whole point of having a client is that it's long lived.  It can better manage connections or have the ability to abort old connections.  It could also potentially create a memory leak if you just keep creating it or at the very least the GC will keep getting hammered.

File: src/github/client/github-client.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (mboudreau) [reply src/github/client/github-client.ts]: Shouldn't we keep it to 60s?  That's usually the default HTTP timeout.

File: src/github/client/interceptors.ts (hunk 8 above)
Comment (mboudreau) [reply src/github/client/interceptors.ts:27]: unless you return `0` as a number flag, which I really doubt would happen

Comment (KAbakumovAtl) [reply src/github/client/interceptors.ts:27]: Nope, timeout can be 0

File: src/sync/pull-request.ts (hunk 9 above)
Comment (KAbakumovAtl) [reply src/sync/pull-request.ts:49]: Forgot to mention. With update of the library the Headers interface become incompatible with the type of the headers from the AxiosResponse.

Do you have any ideas on how to make it better?

File: src/github/client/interceptors.ts (hunk 8 above)
Comment (mboudreau) [reply src/github/client/interceptors.ts:27]: Wouldn't it just return the defaultValue when it's off?

Comment (mboudreau) [reply src/github/client/interceptors.ts:27]: You're also not checking if the config has a value already that's been explicitly set.  I'd probably go with the approach of `config.timeout = config.timeout || await numberFlag(NumberFlags.GITHUB_CLIENT_TIMEOUT, 60000);`

File: src/sync/pull-request.ts (hunk 9 above)
Comment (mboudreau) [reply src/sync/pull-request.ts:49]: Change the `Headers` interface to extend `AxiosResponseHeaders`?

File: src/github/client/interceptors.ts (hunk 8 above)
Comment (KAbakumovAtl) [reply src/github/client/interceptors.ts:27]: No the defaultValue will be returned when LaunchDarkly is offline. Otherwise LD will return the value which we set in LD when targeting is OFF. 

>You're also not checking if the config has a value already that's been explicitly set.
We don't set it explicitly in requests so far. Doesn't make sense to add this check now. We'll get rid of the interceptor as soon as we clean up the flag.

**Review by mboudreau — APPROVED** [review]
